x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,359 @@
+            : shouldPruneTokens ? new TokenPruningConfig()
+            : null);
+        if (queryVectors != null) {
+            return this;
+        } else if (weightedTokensSupplier != null) {
+            TextExpansionResults textExpansionResults = weightedTokensSupplier.get();
+            if (textExpansionResults == null) {
+                return this; // No results yet
+            }
+
+            return new SparseVectorQueryBuilder(
+                fieldName,
+                textExpansionResults.getWeightedTokens(),
+                null,
+                null,
+                shouldPruneTokens,
+                tokenPruningConfig
+            );
+        }
+
+        // TODO move this to xpack core and use inference APIs
+        CoordinatedInferenceAction.Request inferRequest = CoordinatedInferenceAction.Request.forTextInput(
+            inferenceId,
+            List.of(text),
+            TextExpansionConfigUpdate.EMPTY_UPDATE,
+            false,
+            InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+        );
+        inferRequest.setHighPriority(true);
+        inferRequest.setPrefixType(TrainedModelPrefixStrings.PrefixType.SEARCH);
+
+        SetOnce<TextExpansionResults> textExpansionResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            (client, listener) -> executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                CoordinatedInferenceAction.INSTANCE,
+                inferRequest,
+                ActionListener.wrap(inferenceResponse -> {
+
+                    if (inferenceResponse.getInferenceResults().isEmpty()) {
+                        listener.onFailure(new IllegalStateException("inference response contain no results"));
+                        return;
+                    }
+
+                    if (inferenceResponse.getInferenceResults().get(0) instanceof TextExpansionResults textExpansionResults) {
+                        textExpansionResultsSupplier.set(textExpansionResults);
+                        listener.onResponse(null);
+                    } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                        listener.onFailure(new IllegalStateException(warning.getWarning()));
+                    } else {
+                        listener.onFailure(
+                            new IllegalStateException(
+                                "expected a result of type ["
+                                    + TextExpansionResults.NAME
+                                    + "] received ["
+                                    + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                    + "]. Is ["
+                                    + inferenceId
+                                    + "] a compatible model?"
+                            )
+                        );
+                    }
```

Comment:
We should also assert that we only received a single inference result. We are not handling more than one result correctly (yet).